Four new features for the classification website

SimonePiras opened the pull request:
We forked the project to develop four new features:
- [Image histogram](https://github.com/unica-isde/isde-projects-2023-F/issues/1): displays a histogram of the selected image ;
- [Image transformation](https://github.com/unica-isde/isde-projects-2023-F/issues/2): makes the user change some parameters of the selected image before classifying;
- [Download result button](https://github.com/unica-isde/isde-projects-2023-F/issues/3): the user can download the .json file with the results of the classification and the image of the graph representing them;
- [Upload image button](https://github.com/unica-isde/isde-projects-2023-F/issues/4): let the users upload their custom image to be classified.

All of them are solved and are working correctly.

giMannu committed:
image uploaded and classified almost correctly

giMannu committed:
image uploaded and classified almost correctly n°2

giMannu committed:
Handled button aestetic.

giMannu committed:
Uploaded files extension handled.

giMannu committed:
File deletion handled. Still to redirect.

giMannu committed:
Redirecting managed. Still to manage the other buttons.

giMannu committed:
Button appearance further enhanced.

SimonePiras committed:
Download buttons created

giopresenti98 committed:
Created Image Transformation fields in HTML.

SimonePiras committed:
comments on the html button row

SimonePiras committed:
Added download plot button function

alessiatiddia committed:
added histogram_select. it doesn't work yet

alessiatiddia committed:
prints the histogram correctly

alessiatiddia committed:
prints grayscale histogram

SimonePiras committed:
Working download_scores button

SimonePiras committed:
working plot download button

SimonePiras committed:
changed some parameters to make the plot more similar to the frontend one

SimonePiras committed:
Plot error fix

SimonePiras committed:
format

SimonePiras committed:
Merge pull request #5 from SimonePiras/3-download-results-button

3 download results button

giMannu committed:
Merge branch 'main' into 4-upload-image-button

giMannu committed:
Merge pull request #6 from SimonePiras/4-upload-image-button

4 upload image button

alessiatiddia committed:
rgb histograms

alessiatiddia committed:
Merge branch 'main' into 1-image-histogram

alessiatiddia committed:
Merge pull request #7 from SimonePiras/1-image-histogram

1 image histogram

giopresenti98 committed:
Completed the Assignment.

giopresenti98 committed:
Completed the Assignment.

giopresenti98 committed:
Merge branch 'main' into 2-image-transformation

giopresenti98 committed:
Merge pull request #8 from SimonePiras/2-image-transformation

2 image transformation

giopresenti98 committed:
Addressed the download button on upload page issue.

giopresenti98 committed:
Fixed a small bug.

giopresenti98 committed:
Added comments to transformation API

giopresenti98 committed:
Added a new image format and a line of comment that makes the interface clearer to the user.

SimonePiras committed:
Code format edit

SimonePiras committed:
some other format edit

asotgiu commented on app/static/transformationScores.json:
Why is this file versioned?

asotgiu commented on classification_plot.png:
Why is this file versioned?

asotgiu commented on classification_scores.json:
Why is this file versioned?

asotgiu commented on main.py:
```diff
@@ -26,36 +29,265 @@ def info() -> Dict[str, List[str]]:
     image_id = form.image_id
     model_id = form.model_id
     classification_scores = classify_image(model_id=model_id, img_id=image_id)
+    with open('classification_scores.json', 'w') as f:
+        json.dump(classification_scores, f)
```
What happens if multiple users ask for a classification job? The file would be overwritten. Isn't it best to return the results directly? Or, if you reuse them later for other things, another way to do it would be to store the results with a unique name and clean up after the response is returned to the user (or after a while anyway).

asotgiu commented on main.py:
```diff
@@ -26,36 +29,265 @@ def info() -> Dict[str, List[str]]:
+    plt.xlabel('Score')
+    plt.title('Classification Scores')
+
+    # Save the plot as a PNG file
+    plt.savefig('classification_plot.png')
```
What happens if multiple users ask for a classification job? The file would be overwritten. Isn't it best to return the results directly? Or, if you reuse them later for other things, another way to do it would be to store the results with a unique name and clean up after the response is returned to the user (or after a while anyway).

asotgiu commented on main.py:
```diff
@@ -26,36 +29,265 @@ def info() -> Dict[str, List[str]]:
+    request: Request,
+    image_id: UploadFile = File(...)
+):
+    if (
+        image_id.filename.endswith(".jpg") |
+        image_id.filename.endswith(".JPEG") |
+        image_id.filename.endswith(".jpeg") |
+        image_id.filename.endswith(".png") |
+        image_id.filename.endswith(".webp")
+    ):
```
Suggestion (you don't have to change it): it's not safe to check the file format by only looking at its extension. A safer method is to check the magic number, for instance with [python-magic](https://github.com/ahupp/python-magic).

asotgiu commented on main.py:
```diff
@@ -26,36 +29,265 @@ def info() -> Dict[str, List[str]]:
+        image_id.filename.endswith(".png") |
+        image_id.filename.endswith(".webp")
+    ):
+        with open("./app/static/imagenet_subset/n00000000_usersImage.JPEG", "wb") as buffer:
```
What happens if multiple users ask for a classification job? The file would be overwritten.

asotgiu reviewed (changes requested):
All the requested functionalities are correctly implemented, but a few things need to be addressed before approving and merging the PR:

- Please remove files that should not be versioned or provide an explanation on why the file is there (perhaps they should just be added to the .gitignore.
- Please, fix the concurrency problem with the results, the plots, and the uploaded images being stored with the same name. In addition, if a user uploads a file that is not an image, but has a valid extension, the app crashes and the file remains in memory anyway. The error should be handled (for instance with a try-except) and the file should be removed.
- Please, improve the documentation through the code and add the list of resolved issues to the pull request documentation (edit the description of the PR, or at least use the GUI functionality to link the PR to the issues of the target repository).

Request a new review when it's done. There's no need to open a new PR, it's enough to push the new commits to the origin repository (yours) of the PR. In that way, I will see only the new changes and I won't have to review the full code again.
Thanks!

SimonePiras committed:
1. fixed file versioning
2. made so that temporary files are saved with a unique name
3. implemented a function to delete temporary files
4. implemented a try-except to fix the upload crash when uploading the wrong file

SimonePiras committed:
Added documentation to the new functions

Fixed a bug in the upload function

asotgiu commented:
Resolve #1, #2, #3, #4

asotgiu reviewed (approved):
All the issues were solved, well done! I merge the PR to the main repository.
